x/farm/keeper/farm.go

```diff
@@ -0,0 +1,244 @@
+	if !found {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrNotFound, "farm not found")
+	}
+	farm.TotalFarmingAmount = farm.TotalFarmingAmount.Sub(coin.Amount)
+	k.SetFarm(ctx, coin.Denom, farm)
+
+	farmingReserveAddr := types.DeriveFarmingReserveAddress(coin.Denom)
+	if err := k.bankKeeper.SendCoins(ctx, farmingReserveAddr, farmerAddr, sdk.NewCoins(coin)); err != nil {
+		return nil, err
+	}
+
+	// TODO: emit an event
+
+	return withdrawnRewards, nil
+}
+
+func (k Keeper) Harvest(ctx sdk.Context, farmerAddr sdk.AccAddress, denom string) (withdrawnRewards sdk.Coins, err error) {
+	position, found := k.GetPosition(ctx, farmerAddr, denom)
+	if !found {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrNotFound, "position not found")
+	}
+
+	// TODO: prevent incrementing too many periods?
+	withdrawnRewards, err = k.withdrawRewards(ctx, position)
+	if err != nil {
+		return nil, err
+	}
+
+	k.updatePosition(ctx, position)
+
+	// TODO: emit an event
+
+	return withdrawnRewards, nil
+}
+
+// Rewards is a convenient query method for external modules.
+func (k Keeper) Rewards(ctx sdk.Context, farmerAddr sdk.AccAddress, denom string) sdk.DecCoins {
+	position, found := k.GetPosition(ctx, farmerAddr, denom)
+	if !found {
+		return nil
+	}
+	cacheCtx, _ := ctx.CacheContext()
+	endPeriod := k.incrementFarmPeriod(cacheCtx, denom)
+	return k.calculateRewards(cacheCtx, position, endPeriod)
+}
+
+func (k Keeper) calculateRewards(ctx sdk.Context, position types.Position, endPeriod uint64) sdk.DecCoins {
+	if position.StartingBlockHeight == ctx.BlockHeight() {
+		return nil
+	}
+	startPeriod := position.PreviousPeriod
+	return k.rewardsBetweenPeriods(
+		ctx, position.Denom, startPeriod, endPeriod, position.FarmingAmount)
+}
+
+func (k Keeper) initializeFarm(ctx sdk.Context, denom string) types.Farm {
+	farm := types.Farm{
+		TotalFarmingAmount: sdk.ZeroInt(),
+		CurrentRewards:     sdk.DecCoins{},
+		OutstandingRewards: sdk.DecCoins{},
+		Period:             1,
+	}
+	k.SetFarm(ctx, denom, farm)
+	k.SetHistoricalRewards(ctx, denom, 0, types.HistoricalRewards{
+		CumulativeUnitRewards: sdk.DecCoins{},
+		ReferenceCount:        1,
+	})
+	return farm
+}
+
+func (k Keeper) updatePosition(ctx sdk.Context, position types.Position) {
+	farm, found := k.GetFarm(ctx, position.Denom)
+	if !found { // Sanity check
+		panic("farm not found")
+	}
+	prevPeriod := farm.Period - 1
+	k.incrementReferenceCount(ctx, position.Denom, prevPeriod)
+	position.PreviousPeriod = prevPeriod
+	position.StartingBlockHeight = ctx.BlockHeight()
+	k.SetPosition(ctx, position)
+}
+
+func (k Keeper) incrementFarmPeriod(ctx sdk.Context, denom string) (prevPeriod uint64) {
+	farm, found := k.GetFarm(ctx, denom)
+	if !found { // Sanity check
+		panic("farm not found")
+	}
+	unitRewards := sdk.DecCoins{}
+	if farm.TotalFarmingAmount.IsZero() {
+		// TODO: do something special?
+	} else {
+		unitRewards = farm.CurrentRewards.QuoDecTruncate(sdk.NewDecFromInt(farm.TotalFarmingAmount))
+	}
```

Comment:
Nothing special was in my mind. Since we don't allocate rewards to farms with no farmers unlike the `x/distribution` module, I think we can simply use zero `unitRewards` for those farms.